cpp/include/cudf/detail/gather.hpp

```diff
@@ -41,9 +48,9 @@ namespace detail {
  */
 std::unique_ptr<table> gather(table_view const& source_table,
                               column_view const& gather_map,
-                              bool check_bounds                   = false,
-                              bool ignore_out_of_bounds           = false,
-                              bool allow_negative_indices         = false,
+                              bounds check_bounds                 = bounds::NO_CHECK,
+                              out_of_bounds oob                   = out_of_bounds::DONT_IGNORE,
```

Comment:
check_bounds should not have check in the name, but I can't think of a name to suggest :D

Reply:
This is what I hate about using enum class for bool. It makes ridiculous parameter definitions that repeat the same word 3 times and cause you to use annoying abbreviations (oob) or unnecessary underscores.

I actually find these two parameters very confusing. What is the difference between "check_bounds" and "out of bounds"?  Do we need two separate parameters?  How can you not ignore out of bounds indices if you are not checking bounds? WTF?

Not saying this is less clear than the original, but I am saying this needs more care.

Reply:
> It makes ridiculous parameter definitions that repeat the same word 3 times and cause you to use annoying abbreviations (oob) or unnecessary underscores.

True, but it makes the calls readable, which is arguably more important (single declaration, potentially many calls).

Reply:
> I actually find these two parameters very confusing. What is the difference between "check_bounds" and "out of bounds"? Do we need two separate parameters? How can you not ignore out of bounds indices if you are not checking bounds? WTF?

I find these parameters confusing as well but I considered it out of scope for this refactoring. I didn't design the original API. These boolean parameters were already a part of the interface before the scoped enumerators were introduced.